File: infra/lambda/bedrock_agent/journey_search.py
from __future__ import annotations

import gzip
import json
from datetime import datetime, timedelta, timezone
from decimal import Decimal
from typing import Any

import connection_scan_journey_search
import direct_service_journey_search
from dynamodb_analysis import validate_service_date
from request_contract import RequestError


MAX_ROUTE_TIME_MINUTES = 48 * 60
MAX_TRANSFERS = 3
_index_cache: dict[str, tuple[str, dict[str, Any]]] = {}
JST = timezone(timedelta(hours=9))
REALTIME_TOLERANCE = timedelta(minutes=5)


def search(
    s3_client: Any,
    *,
    bucket: str,
    prefix: str,
    snapshot_bucket: str = "",
    snapshot_key: str = "api/traffic/delays.json",
    value: dict[str, Any],
    now: datetime | None = None,
) -> dict[str, Any]:
    request = _validated_request(value)
    operations, realtime = _current_operations(
        s3_client,
        snapshot_bucket,
        snapshot_key,
        request,
        now or datetime.now(timezone.utc),
    )
    delays = {
        train_number: Decimal(str(operation["delayMinutes"]))
        for train_number, operation in operations.items()
    }
    if request["maxTransfers"] <= 1:
        index = _load_index(
            s3_client,
            bucket,
            prefix,
            request["serviceDate"],
            "direct-service-index.json.gz",
            "direct-service-index-v1",
        )
        result = direct_service_journey_search.search_index(
            index,
            delays,
            request,
            operations=operations if realtime["applied"] else None,
            realtime_route_time=realtime.get("snapshotRouteTimeMinutes"),
        )
    else:
        index = _load_index(
            s3_client,
            bucket,
            prefix,
            request["serviceDate"],
            "connection-index.json.gz",
            "timetable-connection-index-v1",
        )
        result = search_index(
            index,
            delays,
            request,
            operations=operations if realtime["applied"] else None,
            realtime_route_time=realtime.get("snapshotRouteTimeMinutes"),
        )
    result["realtime"] = realtime
    result["trace"]["realtime"] = realtime
    print(json.dumps({
        "event": "journey_search_trace",
        "serviceDate": request["serviceDate"],
        "originStation": request["originStation"],
        "destinationStation": request["destinationStation"],
        "trace": result["trace"],
    }, ensure_ascii=False, separators=(",", ":")))
    if not request["includeTrace"]:
        result.pop("trace", None)
    return result


def search_index(
    index: dict[str, Any],
    delays: dict[str, Decimal],
    request: dict[str, Any],
    *,
    operations: dict[str, dict[str, Any]] | None = None,
    realtime_route_time: float | None = None,
) -> dict[str, Any]:
    enriched_request = {
        "transferPace": "standard",
        "rankingPreference": "balanced",
        **request,
    }
    return connection_scan_journey_search.search_index(
        index,
        delays,
        enriched_request,
        operations=operations,
        realtime_route_time=realtime_route_time,
    )


def _validated_request(value: dict[str, Any]) -> dict[str, Any]:
    origin = value.get("originStation")
    destination = value.get("destinationStation")
    service_date = value.get("serviceDate")
    departure = value.get("departureTimeMinutes")
    limit = value.get("limit", 3)
    max_transfers = value.get("maxTransfers", 3)
    transfer_pace = value.get("transferPace", "standard")
    ranking_preference = value.get("rankingPreference", "balanced")
    if not isinstance(origin, str) or not origin.strip():
        raise RequestError(400, "originStationが必要です。")
    if not isinstance(destination, str) or not destination.strip():
        raise RequestError(400, "destinationStationが必要です。")
    if not isinstance(service_date, str):
        raise RequestError(400, "serviceDateが必要です。")
    validate_service_date(service_date)
    if not _number_in_range(departure, 0, MAX_ROUTE_TIME_MINUTES):
        raise RequestError(400, "departureTimeMinutesが不正です。")
    if not isinstance(limit, int) or isinstance(limit, bool) or not 1 <= limit <= 5:
        raise RequestError(400, "limitは1から5にしてください。")
    if (
        not isinstance(max_transfers, int)
        or isinstance(max_transfers, bool)
        or not 0 <= max_transfers <= MAX_TRANSFERS
    ):
        raise RequestError(400, "maxTransfersは0から3にしてください。")
    if transfer_pace not in {"hurried", "standard", "relaxed"}:
        raise RequestError(400, "transferPaceが不正です。")
    if ranking_preference not in {
        "balanced",
        "earliest-arrival",
        "latest-departure",
        "fewest-transfers",
    }:
        raise RequestError(400, "rankingPreferenceが不正です。")
    include_trace = value.get("includeTrace", False)
    if not isinstance(include_trace, bool):
        raise RequestError(400, "includeTraceが不正です。")
    return {
        "originStation": origin.strip(),
        "destinationStation": destination.strip(),
        "serviceDate": service_date,
        "departureTimeMinutes": float(departure),
        "limit": limit,
        "maxTransfers": max_transfers,
        "transferPace": transfer_pace,
        "rankingPreference": ranking_preference,
        "includeTrace": include_trace,
    }


def _load_index(
    s3_client: Any,
    bucket: str,
    prefix: str,
    service_date: str,
    filename: str,
    schema_version: str,
) -> dict[str, Any]:
    key = f"{prefix.strip('/')}/normalized/{service_date}/{filename}"
    try:
        etag = str(s3_client.head_object(Bucket=bucket, Key=key).get("ETag", ""))
        cached = _index_cache.get(key)
        if cached is not None and cached[0] == etag:
            return cached[1]
        _index_cache.clear()
        body = s3_client.get_object(Bucket=bucket, Key=key)["Body"].read()
        value = json.loads(gzip.decompress(body).decode("utf-8"))
    except Exception as error:
        raise RequestError(503, "指定日の検索インデックスを読み込めません。") from error
    if not isinstance(value, dict) or value.get("schema_version") != schema_version:
        raise RequestError(503, "指定日の検索インデックス形式が不正です。")
    _index_cache[key] = (etag, value)
    return value


def _current_operations(
    s3_client: Any,
    bucket: str,
    key: str,
    request: dict[str, Any],
    now: datetime,
) -> tuple[dict[str, dict[str, Any]], dict[str, Any]]:
    current_service_date = (now.astimezone(JST) - timedelta(hours=4)).date().isoformat()
    base = {
        "applied": False,
        "reason": "future-or-past-service-date",
        "currentServiceDate": current_service_date,
    }
    if request["serviceDate"] != current_service_date:
        return {}, base
    if not bucket or not key:
        return {}, {**base, "reason": "snapshot-not-configured"}
    try:
        body = s3_client.get_object(Bucket=bucket, Key=key)["Body"].read()
        snapshot = json.loads(body.decode("utf-8"))
    except Exception:
        return {}, {**base, "reason": "snapshot-unavailable"}
    if not isinstance(snapshot, dict):
        return {}, {**base, "reason": "snapshot-invalid"}
    failed_sources = snapshot.get("failedSources")
    if not isinstance(failed_sources, list) or failed_sources:
        return {}, {**base, "reason": "snapshot-incomplete"}
    try:
        collected_at = datetime.fromisoformat(str(snapshot.get("collectedAt")))
    except ValueError:
        return {}, {**base, "reason": "snapshot-invalid"}
    if collected_at.tzinfo is None or abs(now - collected_at) > REALTIME_TOLERANCE:
        return {}, {**base, "reason": "snapshot-stale"}
    requested_at = (
        datetime.strptime(request["serviceDate"], "%Y-%m-%d").replace(tzinfo=JST)
        + timedelta(minutes=request["departureTimeMinutes"])
    )
    if abs(requested_at - collected_at.astimezone(JST)) > REALTIME_TOLERANCE:
        return {}, {**base, "reason": "search-time-not-current"}
    trains = snapshot.get("trains")
    if not isinstance(trains, dict):
        return {}, {**base, "reason": "snapshot-invalid"}
    operations = {
        train_number: operation
        for train_number, operation in trains.items()
        if isinstance(train_number, str)
        and isinstance(operation, dict)
        and _valid_operation(operation)
    }
    service_start = datetime.strptime(
        request["serviceDate"], "%Y-%m-%d"
    ).replace(tzinfo=JST)
    return operations, {
        **base,
        "applied": True,
        "reason": "current-complete-snapshot",
        "snapshotCollectedAt": collected_at.isoformat(),
        "snapshotRouteTimeMinutes": (
            collected_at.astimezone(JST) - service_start
        ).total_seconds() / 60,
        "operationCount": len(operations),
    }


def _valid_operation(value: dict[str, Any]) -> bool:
    delay = value.get("delayMinutes")
    return (
        isinstance(delay, (int, float))
        and not isinstance(delay, bool)
        and delay >= 0
        and isinstance(value.get("destination"), str)
        and isinstance(value.get("sources"), list)
        and all(isinstance(source, str) for source in value["sources"])
    )


def _number_in_range(value: Any, minimum: float, maximum: float) -> bool:
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and value == value
        and minimum <= value <= maximum
    )

File: backend/agent-api/src/usecases/journey-search.ts
import { searchJourneyIndex, type JourneyOperation, type JourneySearchRequest } from "@raiquora/journey";
import { isValidServiceDate } from "@raiquora/operation";

import { type JsonObject, RequestError } from "../contracts/agent-request.js";
import type { AgentOperation } from "../ports/agent-operation.js";
import type { JourneyDataRepository } from "../ports/journey-data.js";

const constraintNames = ["excludedServiceTypes", "excludedTrainNames", "excludedTrainNumbers", "excludedServiceUids", "requiredServiceTypes", "requiredTrainNames", "requiredTrainNumbers", "allowedServiceTypes"] as const;
export const journeySearchContractVersion = "journey-search-v1";

export function createJourneySearchOperation(repository: JourneyDataRepository, options: { now?: () => Date; log?: (event: string, fields: Record<string, unknown>) => void } = {}): AgentOperation {
  const now = options.now ?? (() => new Date()); const log = options.log ?? (() => undefined);
  return async (value, context) => {
    const request = validatedJourneyRequest(value);
    const realtime = await currentRealtime(repository, request, now());
    const index = await repository.loadIndex(request.serviceDate, request.maxTransfers! <= 1 ? "direct-service" : "connection");
    let result;
    try { result = searchJourneyIndex(request, { index, operations: realtime.operations, realtimeRouteTime: realtime.routeTime }); }
    catch (error) { throw new RequestError(503, error instanceof Error ? error.message : "経路を検索できません。"); }
    const realtimeMetadata = realtime.metadata;
    result.trace.realtime = realtimeMetadata;
    log("journey_search_completed", { requestId: context.requestId, serviceDate: request.serviceDate, journeyCount: result.journeys.length, strategy: result.trace.strategy, realtimeApplied: realtimeMetadata.applied });
    const body: JsonObject = { contractVersion: journeySearchContractVersion, ...result, realtime: realtimeMetadata };
    if (!value.includeTrace) delete body.trace;
    return { body };
  };
}

export function validatedJourneyRequest(value: JsonObject): JourneySearchRequest {
  if (value.contractVersion !== journeySearchContractVersion) throw new RequestError(400, "journey_searchの契約versionが不正です。");
  const originStation = requiredText(value.originStation, "originStationが必要です。");
  const destinationStation = requiredText(value.destinationStation, "destinationStationが必要です。");
  if (typeof value.serviceDate !== "string") throw new RequestError(400, "serviceDateが必要です。");
  if (!isValidServiceDate(value.serviceDate)) throw new RequestError(400, "serviceDateが実在する日付ではありません。");
  if (typeof value.departureTimeMinutes !== "number" || !Number.isFinite(value.departureTimeMinutes) || value.departureTimeMinutes < 0 || value.departureTimeMinutes > 2_880) throw new RequestError(400, "departureTimeMinutesが不正です。");
  const limit = value.limit ?? 3; if (!Number.isInteger(limit) || (limit as number) < 1 || (limit as number) > 5) throw new RequestError(400, "limitは1から5にしてください。");
  const maxTransfers = value.maxTransfers ?? 3; if (!Number.isInteger(maxTransfers) || (maxTransfers as number) < 0 || (maxTransfers as number) > 3) throw new RequestError(400, "maxTransfersは0から3にしてください。");
  const transferPace = value.transferPace ?? "standard"; if (transferPace !== "hurried" && transferPace !== "standard" && transferPace !== "relaxed") throw new RequestError(400, "transferPaceが不正です。");
  const rankingPreference = value.rankingPreference ?? "balanced"; if (rankingPreference !== "balanced" && rankingPreference !== "earliest-arrival" && rankingPreference !== "latest-departure" && rankingPreference !== "fewest-transfers") throw new RequestError(400, "rankingPreferenceが不正です。");
  const constraints = Object.fromEntries(constraintNames.map((name) => [name, constraintList(value[name])]));
  if (["requiredServiceTypes", "requiredTrainNames", "requiredTrainNumbers"].reduce((sum, name) => sum + (constraints[name] as string[]).length, 0) > 4) throw new RequestError(400, "利用したい列車条件が多すぎます。");
  if (value.includeTrace !== undefined && typeof value.includeTrace !== "boolean") throw new RequestError(400, "includeTraceが不正です。");
  return { serviceDate: value.serviceDate, originStation, destinationStation, departureTimeMinutes: value.departureTimeMinutes, limit: limit as number, maxTransfers: maxTransfers as 0 | 1 | 2 | 3, transferPace, rankingPreference, ...constraints };
}

async function currentRealtime(repository: JourneyDataRepository, request: JourneySearchRequest, now: Date): Promise<{ operations?: Record<string, JourneyOperation>; routeTime?: number; metadata: JsonObject }> {
  const currentServiceDate = serviceDateAt(now); const base: JsonObject = { applied: false, reason: "future-or-past-service-date", currentServiceDate };
  if (request.serviceDate !== currentServiceDate) return { metadata: base };
  const snapshot = await repository.loadRealtimeSnapshot(); if (!snapshot) return { metadata: { ...base, reason: "snapshot-unavailable" } };
  if (!Array.isArray(snapshot.failedSources) || snapshot.failedSources.length > 0 || typeof snapshot.collectedAt !== "string" || !isRecord(snapshot.trains)) return { metadata: { ...base, reason: "snapshot-incomplete" } };
  const collected = new Date(snapshot.collectedAt); if (Number.isNaN(collected.getTime()) || Math.abs(now.getTime() - collected.getTime()) > 5 * 60_000) return { metadata: { ...base, reason: "snapshot-stale" } };
  const requestedAt = Date.parse(`${request.serviceDate}T00:00:00+09:00`) + request.departureTimeMinutes * 60_000;
  if (Math.abs(requestedAt - collected.getTime()) > 5 * 60_000) return { metadata: { ...base, reason: "search-time-not-current" } };
  const operations = Object.fromEntries(Object.entries(snapshot.trains).flatMap(([number, operation]) => validOperation(operation) ? [[number, operation]] : [])) as Record<string, JourneyOperation>;
  const routeTime = (collected.getTime() - Date.parse(`${request.serviceDate}T00:00:00+09:00`)) / 60_000;
  return { operations, routeTime, metadata: { ...base, applied: true, reason: "current-complete-snapshot", snapshotCollectedAt: snapshot.collectedAt, snapshotRouteTimeMinutes: routeTime, operationCount: Object.keys(operations).length } };
}

function serviceDateAt(now: Date): string { return new Intl.DateTimeFormat("en-CA", { timeZone: "Asia/Tokyo", year: "numeric", month: "2-digit", day: "2-digit" }).format(new Date(now.getTime() - 4 * 60 * 60_000)); }
function validOperation(value: unknown): value is JourneyOperation { return isRecord(value) && typeof value.delayMinutes === "number" && value.delayMinutes >= 0 && typeof value.destination === "string" && Array.isArray(value.sources) && value.sources.every((source) => typeof source === "string"); }
function requiredText(value: unknown, message: string): string { if (typeof value !== "string" || !value.trim()) throw new RequestError(400, message); return value.trim(); }
function constraintList(value: unknown): string[] { if (value === undefined) return []; if (!Array.isArray(value) || value.length > 8 || value.some((item) => typeof item !== "string" || !item.trim() || item.length > 160)) throw new RequestError(400, "列車の検索条件が不正です。"); return [...new Set(value.map((item) => (item as string).trim()))]; }
function isRecord(value: unknown): value is JsonObject { return typeof value === "object" && value !== null && !Array.isArray(value); }
